### PR TITLE
feat: reset broken sync

### PR DESCRIPTION
**Pull request by SWvheerden**
Description
---
If sync fails resets chain to the highest pow chain the node locally has the data to. 

Motivation and Context
---

See: https://github.com/tari-project/tari/issues/4866

How Has This Been Tested?
---
Unit tests

### CODE REVIEW

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 1]
```diff
@@ -2042,42 +2067,61 @@ fn reorganize_chain<T: BlockchainBackend>(
...
+    Ok(removed_blocks)
+}
+
+fn swap_to_highest_pow_chain<T: BlockchainBackend>(
+    db: &mut T,
+    block_validator: &dyn PostOrphanBodyValidation<T>,
+    chain_strength_comparer: &dyn ChainStrengthComparer,
+) -> Result<(), ChainStorageError> {
+    let metadata = db.fetch_chain_metadata()?;
+    // lets clear out all remaining headers that done have a matching block
```
Comment (stringhandler) [inline base_layer/core/src/chain_storage/blockchain_database.rs]:
```suggestion
    // lets clear out all remaining headers that don't have a matching block
```

**Review by stringhandler — CHANGES_REQUESTED** [review]
I think the logic in add_block should then call this method as well, so that we don't have duplicate code

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 2]
```diff
@@ -1838,75 +1865,104 @@ fn handle_possible_reorg<T: BlockchainBackend>(
...
+    block_validator: &dyn PostOrphanBodyValidation<T>,
+    chain_strength_comparer: &dyn ChainStrengthComparer,
+) -> Result<BlockAddResult, ChainStorageError> {
+    let metadata = db.fetch_chain_metadata()?;
+    // lets clear out all remaining headers that dont have a matching block
+    // rewind to height will first delete the headers, then try delete from blocks, if we call this to the current
+    // height it will only trim the extra headers with no blocks
+    rewind_to_height(db, metadata.height_of_longest_chain())?;
+    let all_orphan_tips = db.fetch_all_orphan_chain_tips()?;
+    if all_orphan_tips.is_empty() {
```
Comment (stringhandler) [inline base_layer/core/src/chain_storage/blockchain_database.rs:1943]: Maybe this check should be before we rewind

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 3]
```diff
@@ -1838,75 +1865,104 @@ fn handle_possible_reorg<T: BlockchainBackend>(
...
+        // we have no orphan chain tips, we have trimmed remaining headers, we are on the best tip we have, so lets
+        // return ok
+        return Ok(BlockAddResult::OrphanBlock);
```
Comment (stringhandler) [inline base_layer/core/src/chain_storage/blockchain_database.rs:1946]: We should probably rename this to `ChainReorgResult`

Comment (stringhandler) [reply base_layer/core/src/chain_storage/blockchain_database.rs:1946]: And `OrphanBlock` to change to `NoChange`

File: base_layer/core/src/chain_storage/blockchain_database.rs (hunk 2 above)
Comment (SWvheerden) [reply base_layer/core/src/chain_storage/blockchain_database.rs:1943]: I specifically did not do this as I use the rewind code here to remove the "extra" headers. 
I want to "main chain" to be "clean" after this run. 

The intended case is this. 

Header sync downloads 10 headers. 
Block sync fails after 5. 
That extra 5 headers will stay on top of the main chain till we remove them again. I want to remove them at this point regardless of whether the new tip is higher than our original one.

**Pull request merged**